File: .agent/skills/transport-data/scripts/generate_e1_stop_times.py
#!/usr/bin/env python3
"""
Generate a JSON file with all stop times for Line E1 (route_id=3).
This will be used as a fallback in the Trip API when RT data is unavailable.
"""
import csv
import json
from pathlib import Path

GTFS_DIR = Path("gtfs_data")
OUTPUT_FILE = Path("public/data/e1_stop_times.json")

def parse_time(time_str):
    """Convert HH:MM:SS to seconds from midnight."""
    parts = time_str.split(':')
    return int(parts[0]) * 3600 + int(parts[1]) * 60 + int(parts[2])

def main():
    # 1. Get all E1 trip IDs (route_id = '3')
    e1_trip_ids = set()
    e1_trip_info = {}
    
    with open(GTFS_DIR / "trips.txt", 'r', encoding='utf-8') as f:
        reader = csv.DictReader(f)
        for row in reader:
            if row['route_id'] == '3':  # Line E1
                e1_trip_ids.add(row['trip_id'])
                e1_trip_info[row['trip_id']] = {
                    'headsign': row.get('trip_headsign', ''),
                    'direction': int(row.get('direction_id', 0))
                }
    
    print(f"Found {len(e1_trip_ids)} E1 trips")
    
    # 2. Get all stop times for E1 trips
    trips_data = {}
    
    with open(GTFS_DIR / "stop_times.txt", 'r', encoding='utf-8') as f:
        reader = csv.DictReader(f)
        for row in reader:
            trip_id = row['trip_id']
            if trip_id not in e1_trip_ids:
                continue
            
            if trip_id not in trips_data:
                trips_data[trip_id] = {
                    'tripId': trip_id,
                    'headsign': e1_trip_info[trip_id]['headsign'],
                    'direction': e1_trip_info[trip_id]['direction'],
                    'stops': []
                }
            
            trips_data[trip_id]['stops'].append({
                'stopId': row['stop_id'],
                'sequence': int(row['stop_sequence']),
                'arrivalTime': parse_time(row['arrival_time']),  # Seconds from midnight
                'departureTime': parse_time(row['departure_time'])
            })
    
    # Sort stops by sequence for each trip
    for trip in trips_data.values():
        trip['stops'].sort(key=lambda s: s['sequence'])
    
    # Convert to list
    result = list(trips_data.values())
    
    print(f"Processed {len(result)} trips with stop times")
    
    # Save to JSON
    OUTPUT_FILE.parent.mkdir(parents=True, exist_ok=True)
    with open(OUTPUT_FILE, 'w', encoding='utf-8') as f:
        json.dump(result, f, ensure_ascii=False)
    
    print(f"Saved to {OUTPUT_FILE}")
    
    # Show sample
    if result:
        sample = result[0]
        print(f"\nSample trip: {sample['tripId']}")
        print(f"  Direction: {sample['direction']}")
        print(f"  Headsign: {sample['headsign']}")
        print(f"  Stops: {len(sample['stops'])}")
        for stop in sample['stops'][:3]:
            h = stop['arrivalTime'] // 3600
            m = (stop['arrivalTime'] % 3600) // 60
            print(f"    {stop['sequence']}: {stop['stopId']} @ {h:02d}:{m:02d}")

if __name__ == '__main__':
    main()

File: scripts/gtfs/generate_e1_stop_times.ts
import fs from 'fs';
import path from 'path';
import { parse } from 'csv-parse/sync';

const GTFS_DIR = 'gtfs_data';
const OUTPUT_FILE = path.join('public', 'data', 'e1_stop_times.json');

function readCsv(fileName: string): any[] {
  const filePath = path.join(GTFS_DIR, fileName);
  if (!fs.existsSync(filePath)) return [];
  const fileContent = fs.readFileSync(filePath, 'utf-8');
  const content = fileContent.charCodeAt(0) === 0xFEFF ? fileContent.slice(1) : fileContent;
  return parse(content, { columns: true, skip_empty_lines: true });
}

function parseTime(timeStr: string): number {
  const parts = timeStr.split(':').map(Number);
  return parts[0] * 3600 + parts[1] * 60 + parts[2];
}

function main() {
  const e1RouteIds = new Set<string>();

  for (const row of readCsv('routes.txt')) {
    if (row.route_short_name === 'E1' || row.route_short_name === '20') {
      e1RouteIds.add(row.route_id);
    }
  }

  if (e1RouteIds.size === 0) {
    console.log("❌ Error: Could not find route E1");
    return;
  }

  console.log(`Found E1 route IDs: ${Array.from(e1RouteIds).join(', ')}`);

  const e1TripIds = new Set<string>();
  const e1TripInfo: Record<string, any> = {};

  for (const row of readCsv('trips.txt')) {
    if (e1RouteIds.has(row.route_id)) {
      e1TripIds.add(row.trip_id);
      e1TripInfo[row.trip_id] = {
        headsign: row.trip_headsign || '',
        direction: parseInt(row.direction_id || '0', 10),
      };
    }
  }

  console.log(`Found ${e1TripIds.size} E1 trips`);

  const tripsData: Record<string, any> = {};

  for (const row of readCsv('stop_times.txt')) {
    const tripId = row.trip_id;
    if (!e1TripIds.has(tripId)) continue;

    if (!tripsData[tripId]) {
      tripsData[tripId] = {
        tripId,
        headsign: e1TripInfo[tripId].headsign,
        direction: e1TripInfo[tripId].direction,
        stops: [],
      };
    }

    tripsData[tripId].stops.push({
      stopId: row.stop_id,
      sequence: parseInt(row.stop_sequence, 10),
      arrivalTime: parseTime(row.arrival_time),
      departureTime: parseTime(row.departure_time),
    });
  }

  for (const trip of Object.values(tripsData)) {
    trip.stops.sort((a: any, b: any) => a.sequence - b.sequence);
  }

  const result = Object.values(tripsData);
  console.log(`Processed ${result.length} trips with stop times`);

  fs.mkdirSync(path.dirname(OUTPUT_FILE), { recursive: true });
  fs.writeFileSync(OUTPUT_FILE, JSON.stringify(result, null, 2), 'utf-8');

  console.log(`✅ Saved to ${OUTPUT_FILE}`);

  if (result.length > 0) {
    const sample = result[0];
    console.log(`\nSample trip: ${sample.tripId}`);
    console.log(`  Direction: ${sample.direction}`);
    console.log(`  Headsign: ${sample.headsign}`);
    console.log(`  Stops: ${sample.stops.length}`);
    for (const stop of sample.stops.slice(0, 3)) {
      const h = Math.floor(stop.arrivalTime / 3600);
      const m = Math.floor((stop.arrivalTime % 3600) / 60);
      console.log(`    ${stop.sequence}: ${stop.stopId} @ ${h.toString().padStart(2, '0')}:${m.toString().padStart(2, '0')}`);
    }
  }
}

main();
